include/tvm/runtime/container.h

```diff
@@ -1410,10 +1344,114 @@ inline String& String::operator=(std::string other) {
 
 inline String& String::operator=(const char* other) { return operator=(std::string(other)); }
 
-inline String operator+(const std::string lhs, const String& rhs) {
-  return lhs + rhs.operator std::string();
+template <typename T, typename U,
+          typename = typename std::enable_if<std::is_same<T, String>::value ||
+                                             std::is_same<T, std::string>::value>::type,
+          typename = typename std::enable_if<std::is_same<U, String>::value ||
+                                             (std::is_same<U, std::string>::value &&
+                                              !std::is_same<T, U>::value)>::type>
+inline String operator+(const T& lhs, const U& rhs) {
+  size_t lhs_size = lhs.size();
+  size_t rhs_size = rhs.size();
+  char* concat = new char[lhs_size + rhs_size + 1];
+  std::memcpy(concat, lhs.data(), lhs_size);
+  std::memcpy(concat + lhs_size, rhs.data(), rhs_size);
+  auto ptr = make_object<StringObj>();
```

Comment:
We will need to have a customized deleter. Directly allocating will cause the memory not being released. Let us use StringObj::FromStd for now. Alternatively, create another subclass of StringObj that contains the customized deleter(hopefully via inplaceArray)

Reply:
We should
1) mark `make_object<StringObj>` as deleted, in case there is any misuse in the codebase
2) have a separate subclass of StringObj to make it clearer

Reply:
In my implementation of runtime::Array, I overload `make_object<ArrayNode>` to `make_inplace_array`; In Map, I dispatch `make_object<MapNode>` to `make_object<SmallMapNode>`

Reply:
There is no dispatch needed because all operations only differs in creation time

Reply:
yes, you are right.

Reply:
we can probably have a `StringObj::FromCharPtr` for `const char*` and make the constructor use `make_inplace_array`. Currently we just use `std::string` for the `const char*`.

Reply:
Let us use FromStd for now. The customized deleter is an incremental change by changing the constructor of `String(const char*)` which needs some extra efforts